homeassistant/components/template/switch.py

```diff
@@ -58,19 +60,22 @@
         state_template = device_config[CONF_VALUE_TEMPLATE]
         icon_template = device_config.get(CONF_ICON_TEMPLATE)
         entity_picture_template = device_config.get(CONF_ENTITY_PICTURE_TEMPLATE)
+        availability_template = device_config.get(CONF_AVAILABILITY_TEMPLATE)
         on_action = device_config[ON_ACTION]
         off_action = device_config[OFF_ACTION]
-        entity_ids = (
+        entity_ids = set(
             device_config.get(ATTR_ENTITY_ID) or state_template.extract_entities()
         )
-
         state_template.hass = hass
 
-        if icon_template is not None:
-            icon_template.hass = hass
+        templates = [icon_template, entity_picture_template, availability_template]
 
-        if entity_picture_template is not None:
-            entity_picture_template.hass = hass
+        for template in templates:
+            if template is not None:
+                template.hass = hass
+                temp_entities = template.extract_entities()
+                if temp_entities is not None:
+                    entity_ids |= set(temp_entities)
```

Comment:
I'm not sure if we should touch the entity_ids list and work like with the sensor template

Reply:
@pvizeli , trying to understand. 

I think what you're saying is:
- If the config `entity_id` is set us it and ignore any entities sin the templates
- otherwise build a list of entities in the templates and use them
- but if any those templates `extract_entities()` return `*` (MATCH_ALL), then use the `MATCH_ALL` as the list of entities (and print a warning message)

IF that is the case, I have already been working on a refactor to extract this behaviour to a helper for the module and use it in all 8 template components.

here is the gist: https://gist.github.com/grillp/66d50d7a93c96aa7f02472ce8538ade0

I was waiting for all 8 Template PRs to go through before submitting it.

Reply:
I have fixed the entity extraction in line with the refactor above to bring it into line with the `sensor` entity extraction